src/ModelContextProtocol.Core/Protocol/Reference.cs

```diff
@@ -35,55 +34,154 @@ public class Reference
+    {
+        /// <inheritdoc/>
+        public override Reference? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
 
-    /// <summary>
-    /// Gets or sets the name of the prompt or prompt template.
-    /// </summary>
-    [JsonPropertyName("name")]
-    public string? Name { get; set; }
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new JsonException();
+            }
 
-    /// <inheritdoc />
-    public override string ToString() => $"\"{Type}\": \"{Uri ?? Name}\"";
+            string? type = null;
+            string? name = null;
+            string? title = null;
+            string? uri = null;
 
-    /// <summary>
-    /// Validates the reference object to ensure it contains the required properties for its type.
-    /// </summary>
-    /// <param name="validationMessage">When this method returns false, contains a message explaining why validation failed; otherwise, null.</param>
-    /// <returns>True if the reference is valid; otherwise, false.</returns>
-    /// <remarks>
-    /// For "ref/resource" type, the <see cref="Uri"/> property must not be null or empty.
-    /// For "ref/prompt" type, the <see cref="Name"/> property must not be null or empty.
-    /// </remarks>
-    public bool Validate([NotNullWhen(false)] out string? validationMessage)
-    {
-        switch (Type)
-        {
-            case "ref/resource":
-                if (string.IsNullOrEmpty(Uri))
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType != JsonTokenType.PropertyName)
                 {
-                    validationMessage = "Uri is required for ref/resource";
-                    return false;
+                    continue;
                 }
-                break;
 
-            case "ref/prompt":
-                if (string.IsNullOrEmpty(Name))
+                string? propertyName = reader.GetString();
+                bool success = reader.Read();
+                Debug.Assert(success, "STJ must have buffered the entire object for us.");
+
+                switch (propertyName)
                 {
-                    validationMessage = "Name is required for ref/prompt";
-                    return false;
+                    case "type":
+                        type = reader.GetString();
+                        break;
+
+                    case "name":
+                        name = reader.GetString();
+                        break;
+
+                    case "uri":
+                        uri = reader.GetString();
+                        break;
+
+                    default:
+                        break;
                 }
-                break;
+            }
+
+            switch (type)
```

Comment:
I take it this is a workaround for `AllowOutOfOrderMetadataProperties` not being available? If so, could you add a TODO on the converter?

Reply:
Yes